pkg/test/kubeharness.go

```diff
@@ -0,0 +1,272 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package test
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"os"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/kubebuilder-declarative-pattern/mockkubeapiserver"
+
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/dynamic"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/crd/crdloader"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/logging"
+)
+
+// KubeHarness is a test harness that brings up a kube-apiserver (only).
+type KubeHarness struct {
+	*testing.T
+	Ctx context.Context
+
+	client     client.Client
+	restConfig *rest.Config
+
+	KubeEvents *MemoryEventSink
+}
+
+// NewKubeHarness creates a new KubeHarness.
+func NewKubeHarness(ctx context.Context, t *testing.T) *KubeHarness {
+	ctx, ctxCancel := context.WithCancel(ctx)
+	t.Cleanup(func() {
+		ctxCancel()
+	})
+	log := log.FromContext(ctx)
+
+	h := &KubeHarness{
+		T:   t,
+		Ctx: ctx,
+	}
+
+	loadCRDs := true
+	if targetKube := os.Getenv("E2E_KUBE_TARGET"); targetKube == "envtest" || targetKube == "" {
+		env := &envtest.Environment{
+			ControlPlaneStartTimeout: time.Minute,
+			ControlPlaneStopTimeout:  time.Minute,
+		}
+
+		h.Logf("starting envtest apiserver")
+		restConfig, err := env.Start()
+		if err != nil {
+			h.Fatalf("error starting test environment: %v", err)
+		}
+
+		t.Cleanup(func() {
+			if err := env.Stop(); err != nil {
+				h.Errorf("error stopping envtest environment: %v", err)
+			}
+		})
+
+		h.restConfig = restConfig
+	} else if targetKube := os.Getenv("E2E_KUBE_TARGET"); targetKube == "mock" {
```

Comment:
Shall this be `E2E_GCP_TARGET`?

Reply:
I don't believe so, this is to enable (fast) mock-kubeapiserver instead of (slow) envtest

Reply:
ohoh, I see. Both envs use value "mock". 🤣